[CORE-6836] schema registry json internal references

andijcr opened the pull request:
Note: in draft since I need to port tests and reference merging

Support for internal references for json schema compat checks

# Implementation strategy:

## prepoc phase:

unbundle schemas and save them in a map <$id value, <jsonpointer to schema, dialect>>

note: it’s possible to have nested bundled schemas, the code eagerly collects anything that looks like a bundled schema.

then 
for each schema (main or bundled), resolve all relative $ref to absolute, using the correct base uri.

this is done by manually traversing the tree in parse_json, looking for "$id" or "id", and "$ref".

The object is then modified in memory.

No error is thrown in this phase, as it is relevant only if a reference is accessed during the compat check phase

## Compat check phase:

Hook in get_schema to iteratively resolve references and use the previous map to handle references into bundled schemas.

Error out if:
- the local ref is invalid
- the absolute schema uri does not exist
- The maximum depth of references is reached. This is to safeguard against direct recursive references and indirect recursive references

To safeguard against recursion, a counter in schema_context is decremented every time a ref is resolved, and this counter is shared with all children of `is_superset`, the chosen value is 20

get_object_or_empty make use of get_schema but it does not modify the counter for its siblings, need to think about this.

# not implemented yet

- reference merging: need to port the code
- different dialects for bundled schema: need to keep track of the dialect for each json::Value
- external references: part of another jira ticket.


Fixes https://redpandadata.atlassian.net/browse/CORE-6836

## Backports Required

<!-- Checking at least one of the checkboxes is REQUIRED if this PR is not a backport. -->

- [ ] none - not a bug fix
- [ ] none - this is a backport
- [ ] none - issue does not exist in previous branches
- [ ] none - papercut/not impactful enough to backport
- [ ] v24.2.x
- [ ] v24.1.x
- [ ] v23.3.x

## Release Notes

<!--
If the changes in this PR do not need to be mentioned in the release
notes, then don't add a sub-section and simply list `none`, e.g.

* none

Otherwise, adding a sub-section or `none` is REQUIRED if the PR is not a backport PR.
If this is a backport PR, adding contents to this section will override
the release notes section inherited from the original PR to dev.

Add one or more of the sub-sections with a short description bullet
point of the change, e.g.

### Bug Fixes

* Short description of the bug fix if this is a PR to `dev` branch.

### Features

* Short description of the feature. Explain how to configure.

### Improvements

* Short description of how this PR improves existing behavior.

-->

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -253,12 +253,27 @@ struct pjp {
+        return &(it->second);
+    }
+
+    int ref_max_allowed_depth() const { return _ref_units; }
+    void decrease_ref_units() { --_ref_units; }
```
It might be cleaner to name them:
```suggestion
    int remaining_ref_units() const { return _ref_units; }
    int consume_ref_unit() { return --ref_units; }

andijcr replied:
much better thanks

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -618,11 +618,127 @@ json_type_list normalized_type(const json::Value& v) {
+    auto [id_uri, fragment_p] = get_uri_fragment(ref_it->value.GetString());
+
+    // resolve the reference:
+    for (auto allowed_jumps = ctx.ref_max_allowed_depth(); allowed_jumps > 0;
+         --allowed_jumps, ctx.decrease_ref_units()) {
```
I guess there are options, but perhaps:
```suggestion
    while (ctx.consume_ref_unit() != 0) {

andijcr replied:
much better thanks

pgellert reviewed (commented):
great effort, looks good so far. I'd love to see some tests for all this behaviour and various edge cases 😄

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -205,6 +244,14 @@ struct pj {
+struct pjp {
+    const json::Pointer& p;
+    friend std::ostream& operator<<(std::ostream& os, const pjp& p) {
+        p.p.Stringify(os);
```
:+1:

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -380,18 +385,20 @@ result<document_context> parse_json(iobuf buf) {
-            if (it->value().is_string() == false || dialect == std::nullopt) {
+            if (
+              it->value().is_string() == false
+              || maybe_dialect == std::nullopt) {
```
:+1:

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+        return from_uri(dialect_it->value().as_string_view());
+    }();
+
+    auto maybe_draft4_id_it = this_obj.find("id");
+    auto maybe_id_it = this_obj.find("$id");
```
i tried to come with a simpler form but the fact is that we know that we are in bundled schema if we find a "id" member, and then if we are in a bundled schema then "$schema" can be analyzed, and the dialect has agree to "$id" or "id. it's kinda a chicken/egg problem,
so that's why i just try to find both, and if at least one is found then i assume we are in bundled schema and i can check if it's correct or not.

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+        auto tmp_base_uri = jsoncons::uri{id_it->value().as_string()};
+        if (!tmp_base_uri.is_absolute()) {
+            // base_uri can be relative to the parent schema
+            tmp_base_uri.resolve(base_uri);
```
fixed thanks

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+    for (auto& e : this_obj.object_range()) {
+        if (!e.value().is_object()) {
+            continue;
+        }
```
fixed

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+    // accessed by a $ref. so we don't register them in the base_uris_map, an
+    // will raise an error later if we try to access them. keyword | dialect  |
+    // is bundled schema
+    //   "$id" |    any   |       yes
```
:+1:

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -107,6 +109,43 @@ constexpr std::optional<json_schema_dialect> from_uri(std::string_view uri) {
+// present, then the default value "" is used
+using id_to_schema_pointer = absl::flat_hash_map<
+  json_id_uri,
+  std::pair<json::Pointer, json_schema_dialect>,
```
waiting to get to external refs in the next pr, to see if it makes the code clearer once i factor those in.

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+    //   "$id" |  draft4  |       no
+    //   "id"  |    any   |       yes
+    //   "id"  | >=draft6 |       no
+    //   "id"  |  draft4  |       yes
```
good spot, i fixed the table, hopefully it should be clearer now

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+        auto ref_uri = jsoncons::uri{ref_it->value().as_string()};
+        if (!ref_uri.is_absolute()) {
+            ref_it->value() = ref_uri.resolve(base_uri).string();
+        }
```
i simplified the block , resolve internally is a no-op if `this` is already absolute

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1956,6 +1974,169 @@ void sort(json::Value& val) {
+            return;
+        }
+
+        // run validation if we are not a guaranteed to be in proper schema
```
it is, $def` and `definitions` are already schema-validated, and i think the 99% use case is to use them for bundled schemas

andijcr replied:
Revisiting this: I was checking the meta schema validation and realized that this check is too brittle and we can't rely on this heuristic, so i just removed it and now validation is always performed

andijcr commented:
force push: addressed comments, added test cases

vbotbuildovich commented:
<!--NEW FAILURES-->
new failures in https://buildkite.com/redpanda/redpanda/builds/55458#01924419-c191-4958-bfe6-605084731cb1:
```
"rptest.tests.schema_registry_test.SchemaRegistryAutoAuthTest.test_normalize.dataset_type=JSON"
"rptest.tests.schema_registry_test.SchemaRegistryTest.test_normalize.dataset_type=JSON"
```
<!--38618ffd-2503-49fe-a4b0-9a5127b914a6-->

andijcr commented:
force push: rebase on dev, switched from jsoncons::json to jsoncons::ojson to preserve insertion order of the json keys.

```
    the switch to jsoncons::json in parse_json means that the function always performed
    key sorting, making the "normalize" flag redundant

    jsoncons can work in insertion-order mode, to do so we switch to the
    type alias jsoncons::ojson.

    this is done to preserve the original order of the input
    see
    tests/rptest/tests/schema_registry_test.py::SchemaRegistryAutoAuthTest.test_normalize
    for an example where this can be observed externally from the schema
    registry API

```

andijcr commented:
force push: removed brittle heuristic, fixed a warning for an error message, tried to appease clang-format

andijcr commented:
https://github.com/redpanda-data/redpanda/pull/23461#issuecomment-2388065178
edit: issue was between keyboard and chair

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -602,11 +602,116 @@ json_type_list normalized_type(const json::Value& v) {
+          error_code::schema_invalid,
+          fmt::format(
+            "Unsupported merging of references with base object: '{}'",
+            pj{candidate})});
```
This will give a subset of the doc, but I wonder if the `$ref` is more helpful?

andijcr replied:
removed in the following commit set (kept here to support bisecting)

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -137,6 +137,37 @@ struct document_context {
+      : storage(std::move(d)) {}
+
+    operator const json::Value&() const {
+        return std::visit(
+          ss::make_visitor(
```
note: `ss::visit` should return `decltype(auto)`

andijcr replied:
yes, I always forget that but that's the reason why I default to std::visit(ss::make_visitor (that and the support for multiple variants)

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -670,6 +670,94 @@ resolve_pointer(const json::Pointer& p, const json::Value& root) {
+        for (auto& [k, v] : obj) {
+            if (k != ref_key) {
+                doc.AddMember(
+                  json::Value(k, alloc), json::Value(v, alloc), alloc);
```
note: I guess we don't need to copy any strings (`json::Value` 3rd param) since everything is kept alive in another doc somewhere.

andijcr replied:
i tried but that's not how the rapidjson api works, the Document has to own <key,value>

BenPope replied:
> i tried but that's not how the rapidjson api works, the Document has to own <key,value>

I thought the default was to not copy?

andijcr replied:
https://github.com/Tencent/rapidjson/blob/815e6e7e7e14be44a6c15d9aefed232ff064cad0/include/rapidjson/document.h#L1381C9-L1387C36

basically it requires lvalues `Value::AddMember(Value&, Value&, Allocator&)` but the implementation will take ownership and set the inputs as null.
i haven't found another API to do it

BenPope replied:
> https://github.com/Tencent/rapidjson/blob/815e6e7e7e14be44a6c15d9aefed232ff064cad0/include/rapidjson/document.h#L1381C9-L1387C36
> 
> basically it requires lvalues `Value::AddMember(Value&, Value&, Allocator&)` but the implementation will take ownership and set the inputs as null. i haven't found another API to do it

I was thinking of this: https://github.com/Tencent/rapidjson/blob/815e6e7e7e14be44a6c15d9aefed232ff064cad0/include/rapidjson/document.h#L742

andijcr replied:
ah, you mean `bool copyConstStrings = false)`
do we dare? seems doable

BenPope replied:
> ah, you mean `bool copyConstStrings = false)` do we dare? seems doable

Nothing to be done, I think.

andijcr replied:
ok i added it

BenPope replied:
> ok i added it

OK, but it was still the default! Explicit is fine.

andijcr replied:
🤦 i wrote `false` in the code instead of `true`

ok i'll change it in the next PR

BenPope replied:
> 🤦 i wrote `false` in the code instead of `true`
> 
> ok i'll change it in the next PR

I'm not asking you to change anything.  We're already NOT copying, that's the default. That seems more efficient. The first comment was ensuring that it's safe to take by reference, since the other one is still alive. You added comments to that effect.

BenPope commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -670,6 +670,94 @@ resolve_pointer(const json::Pointer& p, const json::Value& root) {
+
+    // append everything to the allof array
+    auto& res_alloc = res->GetAllocator();
+    auto& allof_array = res->FindMember("allOf")->value;
```
nit: Maybe `allof_array.Reserve()`?

andijcr replied:
done

BenPope reviewed (commented):
Looks good. Not reviewed all the tests yet.

vbotbuildovich commented:
<!--RETRIED FAILURES-->
ducktape was retried in https://buildkite.com/redpanda/redpanda/builds/55641#01924cdc-6216-4f22-bddd-a8d4fadad868
<!--cbaaffcd-fad4-4ff2-bb05-d74b11cfdbeb-->

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -345,56 +345,64 @@ try_validate_json_schema(const jsoncons::json& schema) {
-    return {std::move(schema), validation_res.assume_value()};
+    auto schema_stream = json::chunked_input_stream{std::move(buf)};
+    auto rapidjson_schema = json::Document{};
+    rapidjson_schema.ParseStream(schema_stream);
```
added it back

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1941,6 +1958,155 @@ void sort(json::Value& val) {
+        if (value.is_object()) {
+            collect_and_fix(key, value);
+        } else if (value.is_array()) {
+            for (auto i = 0u; i < value.size(); ++i) {
+                if (value[i].is_object()) {
+                    collect_and_fix(i, value[i]);
+                }
+            }
```
😭 in the interest of getting this pr in, I'll rework this in the next pr for external refs

andijcr replied:
on a second thought, I don't think it's possible meaningfully to have arrays of arrays with some reachable references: 
"type": "array" requires prefixItems to be a schema of objects, 
similarly allOf/oneOf/anyOf are schema arrays

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -238,12 +238,27 @@ struct pjp {
 
 private:
     const json_schema_definition::impl& _schema;
+    int _ref_units;
```
done

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1941,6 +1958,155 @@ void sort(json::Value& val) {
+    // entry point to collect all bundled schemas
+    auto bundled_schemas = id_to_schema_pointer{};
+    if (doc.is_bool()) {
+        bundled_schemas.insert_or_assign(
+          {}, std::pair{json::Pointer{}, dialect});
```
i wanted to keep the invariant of bundled schemas always having at least an entry for the root. 
in this specific case there is no practical purpose, but since in the rest of the file we try to treat `true` as `{}` i felt that it's no harm to extend this also here.

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1904,6 +2189,154 @@ void sort(json::Value& val) {
+    // but do not generate errors here, since they are problematic only when a
+    // accessed by a $ref. so we don't register them in the base_uris_map, an
+    // will raise an error later if we try to access them.
+    // keyword | $schema  | is bundled schema
+    //   "$id" |          | if root is >=draft6
+    //   "$id" | >=draft6 |       yes
+    //   "$id" |  draft4  |       no
+    //   "id"  |          | if root is draft4
+    //   "id"  | >=draft6 |       no
+    //   "id"  |  draft4  |       yes
```
Also, I know that draft5 exists but it might be simpler to just write `draft4` and `>draft4` (or `!=draft4`) here

andijcr replied:
done

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1904,6 +2189,154 @@ void sort(json::Value& val) {
+    // entry point to collect all bundled schemas
+    auto bundled_schemas = id_to_schema_pointer{};
+    if (doc.is_bool()) {
+        bundled_schemas.insert_or_assign(
+          {}, std::pair{json::Pointer{}, dialect});
```
Would this be equivalent? I'm wondering if (1) there's a difference in behaviour here between an empty object and a bool ("true") and (2) whether the creation of `root_id` could be pulled out of the branches.
```suggestion
        bundled_schemas.insert_or_assign(
          json_id_uri{""}, std::pair{json::Pointer{}, dialect});
```

andijcr replied:
equivalent, but reworked the function to factor out this part as for (2)

pgellert commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -314,58 +390,92 @@ try_validate_json_schema(const jsoncons::json& schema) {
+    auto bundles_schemas_map = collect_bundled_schema_and_fix_refs(
+      schema, dialect);
+
+    // to use rapidjson we need to serialized schema again
+    auto iobuf_os = iobuf_ostream{};
```
```suggestion
    // to use rapidjson we need to serialize the schema again
    // We take a copy of the jsoncons schema here because it has the fixed-up references that we want to use for compatibility checks
    auto iobuf_os = iobuf_ostream{};
```

andijcr replied:
done

pgellert reviewed (commented):
I left a few minor suggestions/questions but it pretty much looks good.

BenPope reviewed (dismissed):
I have 2 [commits here](https://github.com/redpanda-data/redpanda/compare/feat/core-6836/schema-registry-json-references...BenPope:redpanda:feat/core-6836/schema-registry-json-references): 

1. I didn't see tests for relative references so I added some
2. Recursion results in stack-overflow (this should be fixed)

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1904,6 +2189,154 @@ void sort(json::Value& val) {
+            if (dialect_it == this_obj.object_range().end()) {
+                // we can keep using the parent dialect
+                return dialect;
+            }
```
done

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1904,6 +2189,154 @@ void sort(json::Value& val) {
+            return;
+        }
+
+        // run validation since we are not a guaranteed to be in proper schema
+        if (validate_json_schema(maybe_new_dialect.value(), this_obj)
+              .has_error()) {
+            // stop exploring this branch, the schema is invalid
+            return;
+        }
```
checking the ref implementation behavior, will add a commit to match the result

andijcr replied:
taking the stricter approach to throw if the bundled schema is invalid.

in a follow up this can be relaxed if the bundled is never actually referenced

andijcr commented on src/v/pandaproxy/schema_registry/json.cc:
```diff
@@ -1904,6 +2189,154 @@ void sort(json::Value& val) {
+            return from_uri(dialect_it->value().as_string_view());
+        }();
+
+        if (maybe_new_dialect.has_value() == false) {
+            // stop scanning this tree, we might be in a bundled schema but we
+            // don't know the dialect.
+            return;
+        }
```
checking the ref implementation behavior, will add a commit to match the result

andijcr commented:
> I have 2 [commits here](https://github.com/redpanda-data/redpanda/compare/feat/core-6836/schema-registry-json-references...BenPope:redpanda:feat/core-6836/schema-registry-json-references):
> 
>     1. I didn't see tests for relative references so I added some
> 
>     2. Recursion results in stack-overflow (this should be fixed)

1. added them
2. fix for this is to ensure that ref resolution is done only in is_superset. to do so i decoupled get_object_or_empty from get_schema. get_object_or_empty does not need the context because it's used for keyword like "properties" that can't be a reference (added a test for this), or for objects that gets passed directly to is_superset, so there is no need to resolve references

BenPope commented on src/v/pandaproxy/schema_registry/test/test_json_schema.cc:
```diff
@@ -154,6 +154,51 @@ static const auto error_test_cases = std::to_array({
+  "$defs": {
+      "$comment": "dialect is draft-04, but id key is '$id'",
+      "$id": "https://example.com/mismatch_id",
+      "$schema": "http://json-schema.org/draft-04/schema#"
+  }
+}
+)",
+    pps::error_info{
+      pps::error_code::schema_invalid,
+      "bundled schema with mismatched dialect "
+      "'http://json-schema.org/draft-04/schema#' for id key"}},
+  error_test_case{
+    R"(
+{
+  "$comment": "the root schema is valid but the bundled schema is not",
+  "$defs": {
+      "$comment": "dialect is unkown",
+      "$id": "https://example.com/mismatch_id",
+      "$schema": "http://json-schema.org/draft-3000/schema#"
+  }
+}
+)",
+    pps::error_info{
+      pps::error_code::schema_invalid,
+      "bundled schema without a known dialect: "
+      "'http://json-schema.org/draft-3000/schema#'"}},
+  error_test_case{
+    R"(
+{
+  "$comment": "the root schema is valid but the bundled schema is not",
+  "$defs": {
+      "$comment": "schema is invalid",
+      "$id": "https://example.com/mismatch_id",
+      "type": "potato"
+  }
+}
+)",
+    pps::error_info{
+      pps::error_code::schema_invalid,
+      R"(bundled schema is invalid. Invalid json schema: '{"$comment":"schema is invalid","$id":"https://example.com/mismatch_id","type":"potato"}'. Error: '/type: Must be valid against at least one schema, but found no matching schemas')"}},
```
It makes sense to me that these are rejected, but this appears stricter than the reference implementation.

I wonder if referencing a schema in the bundled schema changes that?

andijcr replied:
yes so as far as i tested, only if a reference from the root points to an invalid bundled schema, then the whole schema is rejected.

So if a bundled schema is never referenced, then the schema is not rejected. 

to implement this we need to evolve the whole collect+fix function, since the reference implementation only considers references that can be actually accessed, so this is not not considered
`/randomname/$ref` while this is `/properties/a_property_name/$ref`

BenPope replied:
> yes so as far as i tested, only if a reference from the root points to an invalid bundled schema, then the whole schema is rejected.
> 
> So if a bundled schema is never referenced, then the schema is not rejected.
> 
> to implement this we need to evolve the whole collect+fix function, since the reference implementation only considers references that can be actually accessed, so this is not not considered `/randomname/$ref` while this is `/properties/a_property_name/$ref`

Do you think anything needs changing?

BenPope reviewed (commented):
Looks great
